frame/contracts/src/lib.rs

```diff
@@ -367,6 +369,9 @@ pub trait Trait: frame_system::Trait {
 	/// trie id generator
 	type TrieIdGenerator: TrieIdGenerator<Self::AccountId>;
 
+	/// The means of dispatching the calls to the runtime.
+	type Dispatcher: Dispatcher<<Self as Trait>::Call>;
```

Comment:
See the discussion [there](https://github.com/paritytech/substrate/pull/5130#discussion_r392571834). 

TL;DR: I started it without thinking too much (since we doo the same with all other things) and then I found some benefits, namely using that as an escape hatch. I.e. the chain author can implement their own dispatcher that e.g. filters some calls / does additional bookkeeping and then delegates control to the system module. @athei 's idea was to actually introduce `MainDispatcher` to avoid hard-coding `frame_system::Module` directly but to just delegate to `frame_system::Trait::MainDispatcher`.

Reply:
The reasoning is to give the runtime implementer the opportunity to customize the dispatch process for some Module. With the rewrite this is much safer in a way that dispatchers cannot manually call `Dispatchable::dispatch` but must specify a `RootDispatcher` in their implementation.